examples: support FLUSH_TYPE_VISIBILITY on pmem

yangx-jy opened the pull request:
For server with pmem support in 05-flush-to-persistent, rpma_mr_reg()
should call RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY instead of
RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT if direct-pmem-write parameter
is not specified.

Also remove unneeded initialization of flush_type for client.

Signed-off-by: Xiao Yang <yangx.jy@cn.fujitsu.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/rpma/577)
<!-- Reviewable:end -->

grom72 reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)

___
*[examples/05-flush-to-persistent/server.c, line 159 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM17jFp5Zi-ac58eTTu:-MM17jFp5Zi-ac58eTTv:b-qujcp9) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L159)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	ret = rpma_peer_cfg_get_direct_write_to_pmem(pcfg,
> 			&direct_write_to_pmem);
> 	if (ret)
> 		goto err_pcfg_delete;
> ```
</details>

by default direct_write_to_pmem is set to false in rpma_peer_cfg_new()
no need to set it if USE_LIBPMEM is not enabled
___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MM18GsdAgqM_ho2UmZY:blede67) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
> ```
> is_pmem
> ```
is_pmem says that a particular memory region is located in persistent memory
while direct_write_to_pmem tels if we are able to access PMem directly (e.g. DDIO is off)
we do not want to mixed these two things:
- platform capabilities
- memory region capabilities


<!-- Sent from Reviewable.io -->

yangx-jy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[examples/05-flush-to-persistent/server.c, line 159 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM17jFp5Zi-ac58eTTu:-MMInhVa13ULIDvN4BVE:b-pqvvwn) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L159)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

> ```
> 	ret = rpma_peer_cfg_get_direct_write_to_pmem(pcfg,
> 			&direct_write_to_pmem);
> 	if (ret)
> 		goto err_pcfg_delete;
> ```
by default direct_write_to_pmem is set to false in rpma_peer_cfg_new()
no need to set it if USE_LIBPMEM is not enabled
</blockquote></details>

HI @grom72 

Thanks a lot for your comment.
I just want to get the status of direct_write_to_pmem and take use of it .
Do I miss something?

Best Regards,
Xiao Yang
___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMIzP9q3k4JUbc6YhfC:bdtchee) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

> ```
> is_pmem
> ```
is_pmem says that a particular memory region is located in persistent memory
while direct_write_to_pmem tels if we are able to access PMem directly (e.g. DDIO is off)
we do not want to mixed these two things:
- platform capabilities
- memory region capabilities
</blockquote></details>

Hi @grom72
 
It is reasonable for me.
How about the following change?
```
-                       (is_pmem ? RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT :
+                       ((is_pmem && direct_write_to_pmem) ? RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT :
```
Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)

___
*[examples/05-flush-to-persistent/server.c, line 159 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM17jFp5Zi-ac58eTTu:-MMK2NuT3MUM5BMm70AO:b249827) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L159)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

HI @grom72 

Thanks a lot for your comment.
I just want to get the status of direct_write_to_pmem and take use of it .
Do I miss something?

Best Regards,
Xiao Yang
</blockquote></details>

Hi @yangx-jy 
It is not needed.
It is the client side to decide if flush to persistency is allowed or not (based on direct_write_to_pmem flag)
We do not want to mix this with a particular memory region characteristic.
___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMK18eUBhmN41z8Q1L3:b-77nw71) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72
 
It is reasonable for me.
How about the following change?
-                       (is_pmem ? RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT :
+                       ((is_pmem && direct_write_to_pmem) ? RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT :

Best Regards,
Xiao Yang

</blockquote></details>

Hi @yangx-jy 
 
RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT indicates that the memory region is allocated in persistent memory.

direct_write_to_pmem flag indicates that we are allowed to use the APM method of persistency
if direct_write_to_pmem is false, then we can't use APM, and instead, we need another mechanism to ensure remote persistence.
See example 09.


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMK6eH11up4ejH42nKJ:b6t0lri) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Hi @yangx-jy 
 
RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT indicates that the memory region is allocated in persistent memory.

direct_write_to_pmem flag indicates that we are allowed to use the APM method of persistency
if direct_write_to_pmem is false, then we can't use APM, and instead, we need another mechanism to ensure remote persistence.
See example 09.

</blockquote></details>

Hi @grom72

Thanks for your explanation.

Is RPMA_FLUSH_TYPE_VISIBILITY only supported by DRAM?
In another word, does persistent memory support RPMA_FLUSH_TYPE_VISIBILITY?

Perhaps, I misunderstood the semantics of direct_write_to_pmem before.

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)

___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMKiIayBnkj_TdkH5Ib:b-wbuj4y) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72

Thanks for your explanation.

Is RPMA_FLUSH_TYPE_VISIBILITY only supported by DRAM?
In another word, does persistent memory support RPMA_FLUSH_TYPE_VISIBILITY?

Perhaps, I misunderstood the semantics of direct_write_to_pmem before.

Best Regards,
Xiao Yang
</blockquote></details>

Hi @yangx-jy 
It depends on what type of persistent memory you use.
The proposed API is based on IBTA Spec which distinguishes two types of flush:
- to persistency 
- for visibility
In the case of Intel servers, flush to persistency also ensures visibility but not vice-versa.
This is where direct_write_to_pmem flag is taken into consideration.
If direct_write_to_pmem = false we are not able to ensure persistency of remote persistent memory with librpma API, but we are able to ensure visibility with flush(...,RPMA_FLUSH_TYPE_VISIBILITY, ...).


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

yangx-jy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72, @ldorau, and @yangx-jy)

___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMOTgUO2D8Dap7E7AyN:br2mnu5) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Hi @yangx-jy 
It depends on what type of persistent memory you use.
The proposed API is based on IBTA Spec which distinguishes two types of flush:
- to persistency 
- for visibility
In the case of Intel servers, flush to persistency also ensures visibility but not vice-versa.
This is where direct_write_to_pmem flag is taken into consideration.
If direct_write_to_pmem = false we are not able to ensure persistency of remote persistent memory with librpma API, but we are able to ensure visibility with flush(...,RPMA_FLUSH_TYPE_VISIBILITY, ...).

</blockquote></details>

Hi @grom72 

After reading your detailed explanation, I think that rpma_flush() instead of example-05 should be updated to support flush(...,RPMA_FLUSH_TYPE_VISIBILITY, ...) on remote persistent memory.

Best Regards,
Xiao Yang 


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72 and @yangx-jy)



<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 3 unresolved discussions (waiting on @grom72 and @yangx-jy)

___
*[src/conn.c, line 379 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMOyaKB6CIrLmPkxJCv:-MMOyaKB6CIrLmPkxJCw:b-yxdk0z) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L379)):*
> ```
> 	if (type == RPMA_FLUSH_TYPE_VISIBILITY &&
> 	    0 == (flush_type & RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY)) {
> ```

The `RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_VISIBILITY)`.


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 4 unresolved discussions (waiting on @grom72 and @yangx-jy)

*[a discussion](https://reviewable.io/reviews/pmem/rpma/577#-MMOzDdc-VsP6i4GX1GU:-MMOzDdd9oucwuT0sHVd:bjcbz8p) (no related file):*
The builds fail.

---


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
In order to prevent this error:

```
RPMA_FLUSH_TYPE_PERSISTENT is NOT supported
Nov 18 13:08:55.792994 [21682] *ERROR* conn.c: 380: \
rpma_flush: The remote memory region does not support \
flushing to global visibility
```

you just have to do flush() with `RPMA_FLUSH_TYPE_VISIBILITY` instead of `RPMA_FLUSH_TYPE_PERSISTENT`.

> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577#-:-MMP2VM-7eokNo0HPyfy:b-o2iiaj)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @grom72, @ldorau, and @yangx-jy)

___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMP1SCt5iIL_aYGhBUm:b-o2iiaj) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
> ```C
> 			"The remote memory region does not support flushing to global visibility");
> 		return RPMA_E_NOSUPP;
> 	}
> ```

In order to prevent this error:

```
RPMA_FLUSH_TYPE_PERSISTENT is NOT supported
Nov 18 13:08:55.792994 [21682] *ERROR* conn.c: 380: \
rpma_flush: The remote memory region does not support \
flushing to global visibility
```

you just have to do flush() with `RPMA_FLUSH_TYPE_VISIBILITY` instead of `RPMA_FLUSH_TYPE_PERSISTENT`.


<!-- Sent from Reviewable.io -->

yangx-jy commented:
*[a discussion](https://reviewable.io/reviews/pmem/rpma/577#-MMOzDdc-VsP6i4GX1GU:-MMP3feW0M4xVOkaKPHW:b1kqs4t) (no related file):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

The builds fail.
</blockquote></details>

Sorry, I have updated the PR now.

---


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @grom72, @ldorau, and @yangx-jy)

*[a discussion](https://reviewable.io/reviews/pmem/rpma/577#-MMOzDdc-VsP6i4GX1GU:-MMP3p8M66qA3npntYu3:b-yvcnuu) (no related file):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Sorry, I have updated the PR now.
</blockquote></details>

That's OK now

---


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMP3L8G5ZlY4ksXZcSk:bluiga2) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

In order to prevent this error:

```
RPMA_FLUSH_TYPE_PERSISTENT is NOT supported
Nov 18 13:08:55.792994 [21682] *ERROR* conn.c: 380: \
rpma_flush: The remote memory region does not support \
flushing to global visibility
```

you just have to do flush() with `RPMA_FLUSH_TYPE_VISIBILITY` instead of `RPMA_FLUSH_TYPE_PERSISTENT`.
</blockquote></details>

Hi @ldorau 

Thanks for your quick reply. 

Is it possible to do rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY)  on remote persistent memory which sets RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT and disable direct-pmem-write?

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 3 unresolved discussions (waiting on @grom72 and @yangx-jy)

___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMP4T8d7G9ByV4IGRbT:b78ctxz) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @ldorau 

Thanks for your quick reply. 

Is it possible to do rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY)  on remote persistent memory which sets RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT and disable direct-pmem-write?

Best Regards,
Xiao Yang
</blockquote></details>

Hi, what is the problem you want to solve?

The `RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_VISIBILITY)` and the `RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_PERSISTENT)`.


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMP6Oy4DwevFqxUgV7T:b-o1l8zu) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Hi, what is the problem you want to solve?

The `RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_VISIBILITY)` and the `RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_PERSISTENT)`.


</blockquote></details>

Hi @ldorau 

If we register a remote pmem with RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT and disable the direct write on the remote pmem,
calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem returns RPMA_E_NOSUPP.
Is it expected behavior?  I thought calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem can work.

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMPFascAzwBe4k5PnfC:b-f38nde) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @ldorau 

If we register a remote pmem with RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT and disable the direct write on the remote pmem,
calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem returns RPMA_E_NOSUPP.
Is it expected behavior?  I thought calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem can work.

Best Regards,
Xiao Yang
</blockquote></details>

Perhaps, we only make rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY)  work by registering a remote pmem with RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY.
I didn't understand the semantics of direct_write_to_pmem flag clearly.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 3 unresolved discussions (waiting on @yangx-jy)

___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMP8YJ1DT4zPgwhwcx0:b-aao43) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72 

After reading your detailed explanation, I think that rpma_flush() instead of example-05 should be updated to support flush(...,RPMA_FLUSH_TYPE_VISIBILITY, ...) on remote persistent memory.

Best Regards,
Xiao Yang 
</blockquote></details>

Hi @yangx-jy 
example 05 is to demonstrate flush to persistency.
It intentionally returns an error when it is not possible to ensure remote data persistency via librpma API.
In such a case the remote persistency shall be implemented using the GPSPM method demonstrated in example 09.

The library code shall not be changed.

We could extend example 05 in such a way that both flags will be used for is_pmem == true:

RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT | RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY
___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMPAD_dCJDziJCfCRDU:b-p3ajrz) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @ldorau 

If we register a remote pmem with RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT and disable the direct write on the remote pmem,
calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem returns RPMA_E_NOSUPP.
Is it expected behavior?  I thought calling rpma_flush(RPMA_FLUSH_TYPE_VISIBILITY) on the remote pmem can work.

Best Regards,
Xiao Yang
</blockquote></details>

Hi,
it is platform dependent.
It will work on the Intel platform, but in such a case, both flags shall be used for memory registration.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 3 unresolved discussions (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 3 unresolved discussions (waiting on @ldorau and @yangx-jy)

___
*[src/conn.c, line 379 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMOyaKB6CIrLmPkxJCv:-MMPW-sy8s6SKnrSpof8:b-ozxrvj) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L379)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

> ```
> 	if (type == RPMA_FLUSH_TYPE_VISIBILITY &&
> 	    0 == (flush_type & RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY)) {
> ```

The `RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY` flag has to be set in order to do `flush(RPMA_FLUSH_TYPE_VISIBILITY)`.
</blockquote></details>

No changes to the library are required.


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[examples/05-flush-to-persistent/server.c, line 171 at r1](https://reviewable.io/reviews/pmem/rpma/577#-MM18GsdAgqM_ho2UmZX:-MMPXQ-uCMndFSFzXii2:b-156q3l) ([raw file](https://github.com/pmem/rpma/blob/e454a25d4d9ff704c82537a8e28f4f880012cb2f/examples/05-flush-to-persistent/server.c#L171)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Hi @yangx-jy 
example 05 is to demonstrate flush to persistency.
It intentionally returns an error when it is not possible to ensure remote data persistency via librpma API.
In such a case the remote persistency shall be implemented using the GPSPM method demonstrated in example 09.

The library code shall not be changed.

We could extend example 05 in such a way that both flags will be used for is_pmem == true:

RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT | RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY
</blockquote></details>

Hi @grom72 

Got it, big thanks for your reply.
I will just extend example-05 as you suggested.

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @grom72, @ldorau, and @yangx-jy)

___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMPayPf113a2D7VHXpk:b7l82rq) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Hi,
it is platform dependent.
It will work on the Intel platform, but in such a case, both flags shall be used for memory registration.
</blockquote></details>

Yes, exactly, you can just do:

```
rpma_mr_reg(... RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT | RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY ...);
```


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[src/conn.c, line 383 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMP1SCt5iIL_aYGhBUl:-MMPdQQ61nR76nHm-U7u:b-soo8gr) ([raw file](https://github.com/pmem/rpma/blob/a28613f746c18dcfc8bd144ec5fa7300372420d3/src/conn.c#L383)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Yes, exactly, you can just do:

```
rpma_mr_reg(... RPMA_MR_USAGE_FLUSH_TYPE_PERSISTENT | RPMA_MR_USAGE_FLUSH_TYPE_VISIBILITY ...);
```
</blockquote></details>

Hi @ldorau @grom72 

I really appreciate your detailed explanation. 
I have updated the PR as you suggested.

Thanks again.

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577#-:-MMPuKRg33nuEesDL8Av:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @grom72, @ldorau, and @yangx-jy)



<!-- Sent from Reviewable.io -->

ldorau reviewed (approved):
> Reviewed 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72 and @yangx-jy)



<!-- Sent from Reviewable.io -->

ldorau reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577#-:-MMQ2fGF6sGaWuWzJK5O:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72 and @yangx-jy)



<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72 and @yangx-jy)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @yangx-jy)

___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMQERZZ5BmoppicEQur:b-l7owuy) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
> ```
> printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n");
> ```
printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n");
printf("RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");
___
*[examples/05-flush-to-persistent/client.c, line 231 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQE8Ow6iwRKXGgBu3j:-MMQE8Ow6iwRKXGgBu3k:b-jfhkl7) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L231)):*
> ```
> flush_type = RPMA_FLUSH_TYPE_VISIBILITY;
> ```
redundant code


<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72 and @yangx-jy)

___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMQKwNA2kpDEV_LlWOj:b1x062c) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

> ```
> printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n");
> ```
printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n");
printf("RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");

</blockquote></details>

+1

+ I prefer using a single printf in this case.


<!-- Sent from Reviewable.io -->

yangx-jy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @grom72)

___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMQUAwA4_ceHQGdgGDG:b-9jkuj4) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

+1

+ I prefer using a single printf in this case.
</blockquote></details>

Hi @grom72 @janekmi 
 
+1
It is better to add the hint.

Best Regards,

Xiao Yang
___
*[examples/05-flush-to-persistent/client.c, line 231 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQE8Ow6iwRKXGgBu3j:-MMQUdxK8TC3HcSKwMUv:b-p347k3) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L231)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

> ```
> flush_type = RPMA_FLUSH_TYPE_VISIBILITY;
> ```
redundant code
</blockquote></details>

Hi @grom72 

I have removed the initialization of flush_type so could we make the code readable by keeping it?

Best Regards,
Xiao Yang


<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[examples/05-flush-to-persistent/client.c, line 231 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQE8Ow6iwRKXGgBu3j:-MMQXrtdA3uOYarRUdkK:b-7pgtgz) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L231)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72 

I have removed the initialization of flush_type so could we make the code readable by keeping it?

Best Regards,
Xiao Yang
</blockquote></details>

Sorry for speaking fast.

It seems necessary to keep it because I have removed the initialization of flush_type.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @janekmi and @yangx-jy)

___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMS5nT376uk2stnFhL3:b-x6d883) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72 @janekmi 
 
+1
It is better to add the hint.

Best Regards,

Xiao Yang
</blockquote></details>

printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\nRPMA_FLUSH_TYPE_VISIBILITY is used instead\n");


<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMT7gR-5KOZYK3Icr1Y:b-y4j0sp) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\nRPMA_FLUSH_TYPE_VISIBILITY is used instead\n");
</blockquote></details>

Hi @grom72 

Above format will report > 80 characters error from make style, so use the following format instead: 
printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n"
            "RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @grom72)

___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMUAifRDPv1DGjMY_UC:bspn1pt) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, yangx-jy (Xiao Yang) wrote…</i></summary><blockquote>

Hi @grom72 

Above format will report > 80 characters error from make style, so use the following format instead: 
printf("RPMA_FLUSH_TYPE_PERSISTENT is NOT supported\n"
            "RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");
</blockquote></details>

So please use the following one:

```
printf(
	"RPMA_FLUSH_TYPE_PERSISTENT is NOT supported, RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");
```

It will be OK for cstyle.
And please do not use the `\n` character in the middle.


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @grom72)



<!-- Sent from Reviewable.io -->

yangx-jy commented:
___
*[examples/05-flush-to-persistent/client.c, line 230 at r2](https://reviewable.io/reviews/pmem/rpma/577#-MMQERZZ5BmoppicEQuq:-MMULazLDW4WR1DkBXLI:b-v9k0op) ([raw file](https://github.com/pmem/rpma/blob/ac15f541827c446ef161103515f0b212ade3b93b/examples/05-flush-to-persistent/client.c#L230)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

So please use the following one:

```
printf(
	"RPMA_FLUSH_TYPE_PERSISTENT is NOT supported, RPMA_FLUSH_TYPE_VISIBILITY is used instead\n");
```

It will be OK for cstyle.
And please do not use the `\n` character in the middle.
</blockquote></details>

Hi @ldorau 

Sure, I have updated the PR as you said.

Best Regards,
Xiao Yang 


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577#-:-MMUeWZ55RKkJRFdMGne:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: all files reviewed, 1 unresolved discussion (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

ldorau reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577#-:-MMUiWAE-J4CLJVGkq11:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->

janekmi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/577)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @yangx-jy)



<!-- Sent from Reviewable.io -->
